From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Cache world border

The world border is only initialized once when the level is created.
Lookup from data storage map multiple time is quite unnecessary and expensive.

License: GPL-3.0-only (https://www.gnu.org/licenses/gpl-3.0.html)

Additional term under GNU GPLv3 section 7(b):

Any copy or modified version containing a substantial portion of the
original contributions in this patch must preserve the following
author attribution in the patch header, source code, or another
prominent legal notice accompanying the corresponding source:

Original patch by HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>

diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index b51f90087da46d04df10b41f9f939581adfc26a2..55d1259857eb2a2a728aded004acb49be68db922 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -241,6 +241,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     public final dev.shulker.server.async.tracker.AsyncTracker shulker$asyncTracker = new dev.shulker.server.async.tracker.AsyncTracker(this); // Shulker - Multithreaded tracker
     public boolean hasRidableMoveEvent = false; // Purpur - Ridables
     final List<ServerPlayer> realPlayers; // Leaves - skip
+    private @Nullable WorldBorder worldBorder; // Leaf - Cache world border
 
     @Override
     public @Nullable LevelChunk getChunkIfLoaded(int x, int z) {
@@ -2409,8 +2410,14 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     @Override
     public WorldBorder getWorldBorder() {
+        // Leaf start - Cache world border
+        if (this.worldBorder != null) {
+            return this.worldBorder;
+        }
+        // Leaf end - Cache world border
         WorldBorder worldBorder = this.getDataStorage().computeIfAbsent(WorldBorder.TYPE);
         worldBorder.applyInitialSettings(this.levelData.getGameTime());
+        this.worldBorder = worldBorder; // Leaf - Cache world border
         return worldBorder;
     }
 
